# KHAI PHÁ DỮ LIỆU 
# THUẬT TOÁN APRIORI - KHAI THÁC LUẬT KẾT HỢP

## 1. Cơ sở lý thuyết

### 1.1 Thuật toán Apriori

Thuật toán Apriori được công bố bởi R. Agrawal và R. Srikant vào năm 1994 vì để tìm các tập phổ biến trong một bộ dữ liệu lớn. Tên của thuật toán là Apriori vì nó sử dụng kiến thức đã có từ trước (prior) về các thuộc tính, vật phẩm thường xuyên xuất hiện trong cơ sở dữ liệu. Để cải thiện hiệu quả của việc lọc các mục thường xuyên theo cấp độ, một thuộc tính quan trọng được sử dụng gọi là thuộc tính Apriori giúp giảm phạm vi tìm kiếm của thuật toán.

Thuật toán Apriori có thể làm là nhìn vào quá khứ và khẳng định rằng nếu một việc gì đó xảy ra thì sẽ có tỉ lệ bao nhiêu phần trăm sự việc tiếp theo sẽ xảy ra. Nó giống như nhìn vào quá khứ để dự đoán tương lại vậy, và việc này rất có ích cho các nhà kinh doanh. Ví dụ một siêu thị muốn nghĩ cách sắp xếp các gian hàng một cách hợp lí nhất, họ có thể nhìn vào lịch sử mua hàng và sắp sếp các tập sản phẩm thường được mua cùng nhau vào một chỗ. Hoặc một trang web tin tức muốn giới thiệu cho người dùng các bài viết liên quan đến nhau nhất, cũng có thể áp dụng quy luật tương tự.

### 1.2 Một số khái niệm cơ bản 

- `Hạng mục (item)`: là các mặt hàng (VD: trong siêu thị là thịt, cá, trứng, sữa,...).


- `Tập các hạng mục (itemset)`: là tập các hạng mục (VD: Trong giỏ hàng bao gồm {thịt, cá}).


- `Giao dịch (transaction)` : tập các hạng mục được mua trong một giỏ hàng, lưu kèm với mã giao dịch (TID).


- `Mẫu phổ biến (frequent item)`: là mẫu xuất hiện thường xuyên trong tập dữ liệu ( VD: {giấy ăn, bia} xuất hiện khá nhiều trong các giao dịch ).


- `Tập k-hạng mục (k-itemset)`: các tập chứa k các món hàng.


- `Độ phổ biến (support)`: được tính bằng $supp(X) = \frac{count(X)}{|D|}. X = {B, C} $ là tập các hạng mục, D là cơ sở dữ liệu (CSDL) giao dịch.


- `Tập phổ biến (frequent itemset)`: là tập các hạng mục S (itemset) thỏa mãn độ phổ biến tối thiểu (minsupp – do người dùng xác định như 40% hoặc xuất hiện 5 lần). Nếu $ supp(S) \geq minsupp $ thì S là tập phổ biến.


- `Luật kết hợp (association rule)`: kí hiệu $X \rightarrow Y$, nghĩa là khi X có mặt thì Y cũng có mặt (với xác suất nào đó). 


- `Độ tin cậy (confidence)`: được tính bằng conf(X) = \frac{supp(X+Y)}{supp(X)}

### 1.3 Bài toán khai thác luật kết hợp

Cho độ phổ biến tối thiểu (minsupp) và độ tin cậy tối thiểu (minconf) do người dùng xác định.

Cho tập các hạng mục $I={i_1, i_2, ..., i_m}$ và CSDL giao dịch $D={t_1, t_2, ..., t_n}$, với $t_i={i_{i1}, i_{i2}, ..., i_{ik}}$ và $i_{ij} \in I$.

Bài toán khai thác luật kết hợp là bài toán tìm tất cả các luật dạng $X \rightarrow Y$ (X, Y là tập con của I và X giao Y = {}) thỏa mãn độ phổ biến và độ tin cậy tối thiểu. $supp(X \rightarrow Y) \geq minsupp, conf(X \rightarrow Y) \geq minconf$.

### 1.4 Quy trình khai thác tập phổ biến

Quy trình khai thác luật kết hợp

Bước 1: Tìm tất cả các tập phổ biến (theo ngưỡng minsupp).

Bước 2: Xây dựng luật từ các tập phổ biến

Đối với mỗi tập phổ biến S, tạo ra tất cả các tập con khác rỗng của S.
Đối với mỗi tập con khác rỗng A của S $(|A| < |S|)$. Luật $A \rightarrow (S - A)$ là luật kết hợp cần tìm nếu: $conf (A \rightarrow (S - A)) = \frac{supp(S)}{supp(A)} \geq minconf$ .

### 1.5 Thuật toán tìm tập phổ biến Apriori

Phương pháp:

- Tìm tất cả các tập phổ biến 1- hạng mục (C_1).


- Tạo các tập ứng viên kích thước k-hạng mục (k – candidate itemset) từ các tập phổ biến có kích thước (k-1)-hạng mục. Ví dụ, tạo ứng viên C_2 từ tập phổ biến C_1.


- Kiểm tra độ phổ biến của các ứng viên trên CSDL và loại các ứng viên không phổ biến ta được L_i (i = {1, 2, ..., k}).


- Dừng khi không tạo được tập phổ biến hay tập ứng viên L_k = \{\} hay C_k = \{\}.


Ví dụ cụ thể

<img src="img/apriori_example.png" width="600">

## 2. Sử dụng thư viện `aprori` trong python

### 2.1 Dữ liệu

- File: store_data.csv, 7500 giao dịch trong suốt một tuần tại một cửa hàng bán lẻ của Pháp. 


- Muốn quy tắc chỉ cho những mặt hàng được mua ít nhất 5 lần một ngày hoặc 7 x 5 = 35 lần trong một tuần, 
  vì tập dữ liệu cho khoảng thời gian một tuần. 
  Như vậy độ phổ biến cho các mục đó có thể được tính là 35/7500 = 0,0045. 
  

- Độ tin cậy tối thiểu cho các quy tắc là 20% hoặc 0,2. min_lift=3, min_length=2.

### 2.2 Cài đặt

In [1]:
# Importing the libraries
import numpy as np
import csv
import matplotlib.pyplot as plt
import pandas as pd
from apyori import apriori

In [2]:
# data
store_data = pd.read_csv('store_data.csv', header=None)
store_data.head(5)

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

Thư viện Apriori yêu cầu dữ liệu dưới dạng danh sách. Chuyển đổi pandas sang list.

In [3]:
DataFile = open('store_data.csv', 'r')
DataCaptured = csv.reader(DataFile, delimiter=',')
records = list(DataCaptured)

In [4]:
records[:2]

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs']]

Các tham số trong hàm apriori
- `records`: Danh sách muốn trích xuất quy tắc. 
- `min_support`: Chọn các mục có giá trị hỗ trợ (phổ biến) lớn hơn giá trị được chỉ định bởi tham số.
- `min_confidence`: tham số lọc những quy tắc có độ tin cậy lớn hơn ngưỡng tin cậy được chỉ định bởi tham số.
- `min_lift`: tham số chỉ định giá trị tăng tối thiểu cho các quy tắc được liệt kê ngắn gọn.
- `min_length`: tham số chỉ định số lượng mục tối thiểu mà bạn muốn trong quy tắc của mình.

In [5]:
association_rules = apriori(records, min_support=0.0045, min_confidence=0.2)
association_results = list(association_rules)
print(len(association_results))

447


Như vậy có khoảng 447 luật

Ta sẽ sắp xếp các luật theo độ tin cậy confidence.

In [6]:
association_results[0:5]

[RelationRecord(items=frozenset({'mineral water'}), support=0.23836821757099053, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'mineral water'}), confidence=0.23836821757099053, lift=1.0)]),
 RelationRecord(items=frozenset({'burgers', 'almonds'}), support=0.005199306759098787, ordered_statistics=[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'burgers'}), confidence=0.25490196078431376, lift=2.923577382023146)]),
 RelationRecord(items=frozenset({'almonds', 'chocolate'}), support=0.005999200106652446, ordered_statistics=[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'chocolate'}), confidence=0.29411764705882354, lift=1.7950988369310295)]),
 RelationRecord(items=frozenset({'eggs', 'almonds'}), support=0.006532462338354886, ordered_statistics=[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'eggs'}), confidence=0.3202614379084967, lift=1.7821076007059597)]),
 RelationRecord(it

In [7]:
confidence = []
association = []
for i in range (0, len(association_results)):
    confidence.append(association_results[:len(association_results)][i][2][0][2])
    association.append(list(association_results[:len(association_results)][i][0]))

In [8]:
rank = pd.DataFrame([association,  confidence]).T
rank.columns = ['Association', 'confidence']

In [9]:
# Show top 10 higher lift scores
rank.sort_values('confidence', ascending=False).head(10)

Association confidence
380         [soup, frozen vegetables, mineral water]   0.633333
377    [frozen vegetables, mineral water, olive oil]   0.576471
317            [spaghetti, cooking oil, ground beef]   0.571429
419                      [soup, milk, mineral water]   0.561404
431               [shrimp, mineral water, olive oil]   0.557377
308                 [soup, mineral water, chocolate]   0.552632
316               [eggs, cooking oil, mineral water]   0.545455
344                      [soup, eggs, mineral water]   0.544118
370  [frozen vegetables, ground beef, mineral water]   0.543307
422                    [milk, mineral water, turkey]   0.541176

Như vậy có thể thấy: nhóm rau xanh, nước khoáng, soup và nhóm thứ hai thịt bò xay, dầu ăn, mỳ ý và tôm,... là các loại thực phẩm hay được mua cùng nhau.

## 3. Cài đặt thuật toán

Source code của bài viết được lưu trong file: `Apriori.py`

Cài đặt thư viện

In [10]:
import csv
import itertools
DataFile = open('store_data.csv', 'r')
minsup = 0.0045
minconf = 0.2
f1 = "FItems.txt"
f2 = "Rules.txt"

Hàm tìm các tập phổ biến 1- hạng mục.

In [11]:
def L1():
    '''
    Find frequent 1-itemsets
    '''
    # Get all 1-itemsets in the list items and their counts in the dictionary counts
    # Đếm tất cả số lượng của loại mặt hàng trong các hóa đơn (bản ghi)
    DataCaptured = csv.reader(DataFile, delimiter=',')
    data = list(DataCaptured)
    for e in data:
        e = sorted(e)
    count = {}
    for items in data:
        for item in items:
            if item not in count:
                count[(item)] = 1
            else:
                count[(item)] = count[(item)] + 1
    # 7500: số lượng bản ghi
    # Lấy các ứng viên phổ biến
    count2 = {k: v for k, v in count.items() if v >= minsup * 7500}
    return count2, data

Hàm sinh, Từ tập phổ biến $L_{k-1}$ sinh ra tập ứng viên $C_k$.

In [12]:
def generateCk(Lk_1, flag, data):
    '''
    Generate Ck by joining 2 Lk-1
    '''
    Ck = []

    if flag == 1:
        flag = 0
        for item1 in Lk_1:
            for item2 in Lk_1:
                if item2 > item1:
                    Ck.append((item1, item2))
    else:
        for item in Lk_1:
            k = len(item)
        for item1 in Lk_1:
            for item2 in Lk_1:
                if (item1[:-1] == item2[:-1]) and (item1[-1] != item2[-1]):
                    if item1[-1] > item2[-1]:
                        Ck.append(item2 + (item1[-1],))
                    else:
                        Ck.append(item1 + (item2[-1],))
    L = generateLk(set(Ck), data)
    return L, flag

Từ tập ứng viên $C_k$ sử dụng ngưỡng phổ biến (minsupp) sinh ra tập phổ biến $L_k$.

In [13]:
def generateLk(Ck, data):
    '''
    If item in Ck belongs to a transaction,
    it makes it into list Ct
    Then Ct is thresholded to form L
    '''
    count = {}
    for itemset in Ck:
        # print(itemset)
        for transaction in data:
            if all(e in transaction for e in itemset):
                if itemset not in count:
                    count[itemset] = 1
                else:
                    count[itemset] = count[itemset] + 1
    count2 = {k: v for k, v in count.items() if v >= minsup * 7500}
    return count2

Hàm xây dựng luật từ các tập phổ biến.

In [14]:
def rulegenerator(fitems):
    '''
    Generates association rules from the frequent itemsets
    '''
    counter = 0
    for itemset in fitems.keys():
        if isinstance(itemset, str):
            continue
        length = len(itemset)

        union_support = fitems[tuple(itemset)]
        for i in range(1, length):

            lefts = map(list, itertools.combinations(itemset, i))
            for left in lefts:
                if len(left) == 1:
                    if ''.join(left) in fitems:
                        leftcount = fitems[''.join(left)]
                        conf = union_support / leftcount
                else:
                    if tuple(left) in fitems:
                        leftcount = fitems[tuple(left)]
                        conf = union_support / leftcount
                if conf >= minconf:
                    fo = open(f2, "a+")
                    right = list(itemset[:])
                    for e in left:
                        right.remove(e)
                    fo.write(str(left) + ' (' + str(leftcount) + ')' + ' -> ' + str(right) + ' (' + str(
                        fitems[''.join(right)]) + ')' + ' [' + str(conf) + ']' + '\n')
                    print(str(left) + ' -> ' + str(right) + ' (' + str(conf) + ')')
                    counter = counter + 1
                    fo.close()

### Kết quả: Tập phổ biến, Xây dựng luật từ tập phổ biến

In [15]:
def apriori():
    '''
    The runner function
    '''
    L, data = L1()
    flag = 1
    FreqItems = dict(L)
    while (len(L) != 0):
        fo = open(f1, "a+")
        for k, v in L.items():
            fo.write(str(k) + ' >>> ' + str(v) + '\n\n')
        fo.close()

        L, flag = generateCk(L, flag, data)
        FreqItems.update(L)
    print(FreqItems)
    print("---------------------------------------------------------------------------------------------")
    rulegenerator(FreqItems)

In [16]:
apriori()

{'shrimp': 536, 'almonds': 153, 'avocado': 250, 'vegetables mix': 193, 'green grapes': 68, 'whole weat flour': 70, 'yams': 86, 'cottage cheese': 239, 'energy drink': 200, 'tomato juice': 228, 'low fat yogurt': 574, 'green tea': 991, 'honey': 356, 'salad': 37, 'mineral water': 1788, 'salmon': 319, 'antioxydant juice': 67, 'frozen smoothie': 475, 'spinach': 53, 'olive oil': 494, 'burgers': 654, 'meatballs': 157, 'eggs': 1348, 'turkey': 469, 'milk': 972, 'energy bar': 203, 'whole wheat rice': 439, 'whole wheat pasta': 221, 'french fries': 1282, 'soup': 379, 'light cream': 117, 'shallot': 58, 'frozen vegetables': 715, 'spaghetti': 1306, 'pet food': 49, 'cookies': 603, 'cooking oil': 383, 'champagne': 351, 'chocolate': 1230, 'chicken': 450, 'oil': 173, 'fresh tuna': 167, 'tomatoes': 513, 'black tea': 107, 'extra dark chocolate': 90, 'protein bar': 139, 'red wine': 211, 'pasta': 118, 'pepper': 199, 'shampoo': 37, 'rice': 141, 'sparkling water': 47, 'ham': 203, 'body spray': 86, 'pancakes': 7

['tomato juice'] -> ['spaghetti'] (0.24561403508771928)
['parmesan cheese'] -> ['eggs'] (0.22818791946308725)
['tomato sauce'] -> ['mineral water'] (0.4056603773584906)
['ham'] -> ['french fries'] (0.20689655172413793)
['cottage cheese'] -> ['green tea'] (0.200836820083682)
['hot dogs'] -> ['green tea'] (0.205761316872428)
['cooking oil'] -> ['milk'] (0.2245430809399478)
['protein bar'] -> ['spaghetti'] (0.2733812949640288)
['hot dogs'] -> ['mineral water'] (0.2839506172839506)
['light mayo'] -> ['french fries'] (0.2696078431372549)
['vegetables mix'] -> ['spaghetti'] (0.21761658031088082)
['grated cheese'] -> ['ground beef'] (0.21628498727735368)
['butter'] -> ['eggs'] (0.20353982300884957)
['carrots'] -> ['spaghetti'] (0.30434782608695654)
['frozen vegetables'] -> ['eggs'] (0.22797202797202798)
['vegetables mix'] -> ['french fries'] (0.21761658031088082)
['french wine'] -> ['eggs'] (0.26627218934911245)
['avocado'] -> ['french fries'] (0.24)
['escalope'] -> ['mineral water'] (0.21512

['ground beef'] -> ['eggs'] (0.20352781546811397)
['french wine'] -> ['spaghetti'] (0.3431952662721893)
['pancakes'] -> ['spaghetti'] (0.2650771388499299)
['shrimp'] -> ['mineral water'] (0.3302238805970149)
['red wine'] -> ['mineral water'] (0.3886255924170616)
['chicken'] -> ['spaghetti'] (0.2866666666666667)
['pancakes'] -> ['mineral water'] (0.3548387096774194)
['milk'] -> ['eggs'] (0.23765432098765432)
['fresh bread'] -> ['milk'] (0.21362229102167182)
['muffins'] -> ['french fries'] (0.22099447513812154)
['chicken'] -> ['mineral water'] (0.38)
['cottage cheese'] -> ['mineral water'] (0.301255230125523)
['mineral water'] -> ['spaghetti'] (0.2505592841163311)
['spaghetti'] -> ['mineral water'] (0.3430321592649311)
['pancakes'] -> ['eggs'] (0.22861150070126227)
['salmon'] -> ['mineral water'] (0.4012539184952978)
['pancakes'] -> ['chocolate'] (0.20897615708274894)
['turkey'] -> ['spaghetti'] (0.26439232409381663)
['burgers'] -> ['milk'] (0.20489296636085627)
['escalope'] -> ['chocola

['french fries', 'frozen vegetables'] -> ['mineral water'] (0.32167832167832167)
['low fat yogurt', 'milk'] -> ['mineral water'] (0.36363636363636365)
['low fat yogurt', 'mineral water'] -> ['milk'] (0.2)
['french fries', 'milk'] -> ['spaghetti'] (0.2696629213483146)
['french fries', 'spaghetti'] -> ['milk'] (0.2318840579710145)
['eggs', 'olive oil'] -> ['spaghetti'] (0.4222222222222222)
['olive oil', 'spaghetti'] -> ['eggs'] (0.22093023255813954)
['eggs', 'turkey'] -> ['spaghetti'] (0.2671232876712329)
['spaghetti', 'turkey'] -> ['eggs'] (0.31451612903225806)
['ground beef', 'herb & pepper'] -> ['spaghetti'] (0.4)
['herb & pepper', 'spaghetti'] -> ['ground beef'] (0.39344262295081966)
['chocolate', 'low fat yogurt'] -> ['mineral water'] (0.38738738738738737)
['low fat yogurt', 'mineral water'] -> ['chocolate'] (0.2388888888888889)
['ground beef', 'milk'] -> ['mineral water'] (0.503030303030303)
['ground beef', 'mineral water'] -> ['milk'] (0.2703583061889251)
['milk', 'mineral water']

['olive oil', 'spaghetti'] -> ['pancakes'] (0.22093023255813954)
['pancakes', 'spaghetti'] -> ['olive oil'] (0.20105820105820105)
['milk', 'shrimp'] -> ['spaghetti'] (0.2803030303030303)
['shrimp', 'spaghetti'] -> ['milk'] (0.23270440251572327)
['mineral water', 'salmon'] -> ['spaghetti'] (0.3984375)
['salmon', 'spaghetti'] -> ['mineral water'] (0.504950495049505)
['green tea', 'mineral water'] -> ['spaghetti'] (0.26609442060085836)
['green tea', 'spaghetti'] -> ['mineral water'] (0.31155778894472363)
['cake', 'eggs'] -> ['french fries'] (0.23776223776223776)
['cake', 'french fries'] -> ['eggs'] (0.2537313432835821)
['chocolate', 'frozen smoothie'] -> ['spaghetti'] (0.3125)
['frozen smoothie', 'spaghetti'] -> ['chocolate'] (0.29914529914529914)
['cake', 'chocolate'] -> ['spaghetti'] (0.38235294117647056)
['cake', 'spaghetti'] -> ['chocolate'] (0.2867647058823529)
['burgers', 'milk'] -> ['mineral water'] (0.3880597014925373)
['burgers', 'mineral water'] -> ['milk'] (0.28415300546448086)

['mineral water', 'turkey'] -> ['spaghetti'] (0.3333333333333333)
['spaghetti', 'turkey'] -> ['mineral water'] (0.3870967741935484)
['eggs', 'olive oil'] -> ['mineral water'] (0.4888888888888889)
['mineral water', 'olive oil'] -> ['eggs'] (0.21256038647342995)
['frozen vegetables', 'milk'] -> ['mineral water'] (0.4689265536723164)
['frozen vegetables', 'mineral water'] -> ['milk'] (0.30970149253731344)
['milk', 'mineral water'] -> ['frozen vegetables'] (0.23055555555555557)
['mineral water', 'soup'] -> ['olive oil'] (0.2254335260115607)
['olive oil', 'soup'] -> ['mineral water'] (0.582089552238806)
['cooking oil', 'mineral water'] -> ['pancakes'] (0.23178807947019867)
['cooking oil', 'pancakes'] -> ['mineral water'] (0.5932203389830508)
['eggs', 'herb & pepper'] -> ['mineral water'] (0.425531914893617)
['herb & pepper', 'mineral water'] -> ['eggs'] (0.3125)
['burgers', 'green tea'] -> ['spaghetti'] (0.2748091603053435)
['burgers', 'spaghetti'] -> ['green tea'] (0.2236024844720497)
['gr

Đối chiếu với kết quả sử dụng thư viện, ta cũng được một luật phổ biến gần như tương tự. (File Rules.txt có ['cooking oil', 'ground beef'] (63) -> ['spaghetti'] (1306) [0.5714285714285714], ['frozen vegetables', 'olive oil'] (85) -> ['mineral water'] (1788) [0.5764705882352941], ['frozen vegetables', 'soup'] (60) -> ['mineral water'] (1788) [0.6333333333333333])

# END (Tiếp tục chỉnh sửa)